hal/src/i2c.rs

```diff
@@ -0,0 +1,412 @@
+                    }
+
+                    if bytes.is_empty() {
+                        // 0 byte write
+                        self.i2c.cr2.modify(|_, w| {
+                            w.add10().bit7();
+                            w.sadd().bits((addr << 1) as u16);
+                            w.rd_wrn().write();
+                            w.nbytes().bits(0);
+                            w.reload().completed();
+                            w.autoend().automatic();
+                            w.start().start()
+                        });
+                    } else {
+                        let end = bytes.len() / 0xFF;
+
+                        // Process 255 bytes at a time
+                        for (i, bytes) in bytes.chunks(0xFF).enumerate() {
+                            // Prepare to send `bytes`
+                            self.i2c.cr2.modify(|_, w| {
+                                if i == 0 {
+                                    w.add10().bit7();
+                                    w.sadd().bits((addr << 1) as u16);
+                                    w.rd_wrn().write();
+                                    w.start().start();
+                                }
+                                w.nbytes().bits(bytes.len() as u8);
+                                if i != end {
+                                    w.reload().not_completed()
+                                } else {
+                                    w.reload().completed().autoend().automatic()
+                                }
+                            });
+
+                            for byte in bytes {
+                                // Wait until we are allowed to send data
+                                // (START has been ACKed or last byte went through)
+                                busy_wait!(self.i2c, txis, is_empty);
+
+                                // Put byte on the wire
+                                // NOTE(write): Writes all non-reserved bits.
+                                self.i2c.txdr.write(|w| w.txdata().bits(*byte));
+                            }
+
+                            if i != end {
+                                // Wait until the last transmission is finished
+                                busy_wait!(self.i2c, tcr, is_complete);
+                            }
+                        }
+                    }
+
+                    // automatic STOP
+                    // Wait until the last transmission is finished
+                    busy_wait!(self.i2c, stopf, is_stop);
+
+                    self.i2c.icr.write(|w| w.stopcf().clear());
+
+                    Ok(())
+                }
+            }
+
+            impl<PINS> WriteRead for $I2cX<PINS> {
+                type Error = Error;
+
+                fn write_read(&mut self, addr: u8, bytes: &[u8], buffer: &mut [u8]) -> Result<(), Error> {
+                    assert!(!bytes.is_empty() && !buffer.is_empty());
```

Comment:
This one seems like it should be an assert (?)

Add documentation for the panic condition.